### PR TITLE
feat: WAL recovery

### PR DESCRIPTION
**Pull request by pepyakin**
This implements WAL recovery.

### CODE REVIEW

**Comment by pepyakin** [comment]
* **#343** <a href="https://app.graphite.dev/github/pr/thrumdev/nomt/343?utm_source=stack-comment-icon" target="_blank">Graphite</a> 👈
* `master`

This stack of pull requests is managed by Graphite. <a href="https://stacking.dev/?utm_source=stack-comment">Learn more about stacking.</a>
<h2></h2>

 Join @pepyakin and the rest of your teammates on <a href="https://graphite.dev?utm-source=stack-comment">Graphite <b>Graphite</b></a>

File: nomt/src/bitbox/wal/mod.rs [hunk 1]
```diff
@@ -147,3 +152,83 @@ impl WalBlobBuilder {
...
+
+pub struct WalBlobReader<'a> {
+    wal_fd: &'a std::fs::File,
+}
+impl<'a> WalBlobReader<'a> {
+    /// Creates a new WAL blob reader.
+    ///
+    /// The `wal_fd` is expected to be positioned at the start of the WAL file. The file must be
+    /// a multiple of the page size.
+    pub(crate) fn new(wal_fd: &'a std::fs::File) -> anyhow::Result<Self> {
```
Comment (rphmeier) [inline nomt/src/bitbox/wal/mod.rs:183]: nit: full path appears twice in this file. maybe import it?

File: nomt/src/bitbox/mod.rs [hunk 2]
```diff
@@ -142,6 +137,87 @@ impl DB {
...
+                    changed_meta_page_ixs.insert(meta_map.page_index(bucket as usize));
+                }
+
+                // Apply the diff to the page in the ht file.
+                //
+                // The algorithm is:
+                // - read the bucket page from the ht file.
+                // - for each index of a bit in a diff that equals to 1, copy the changed node into
+                //   the page.
+                // - store the changed page.
```
Comment (rphmeier) [inline nomt/src/bitbox/mod.rs:186]: worth expanding a comment on how our assumptions ensure this recovery is clean?

**Comment by rphmeier** [comment]
### Merge activity

* **Aug 27, 11:09 AM EDT**: @rphmeier started a stack merge that includes this pull request via [Graphite](https://app.graphite.dev/github/pr/thrumdev/nomt/343).

**Pull request merged**